### ISSUE TITLE
 enw_aggregate_cumulative(timestep = "week") not working as expected with data that is reported weekly but is daily

### ISSUE THREAD

**Issue by natemcintosh** [epinowcast/epinowcast#335]
**Describe the bug**
I think  `enw_aggregate_cumulative(timestep = "week")` is either dropping values, or shifting values to a different week.

**To Reproduce**
```r
library(epinowcast)
library(data.table)

show_bug <- function() {
  # This is a chunk of data produced after running enw_filter_reference_dates
  # and enw_filter_report_dates
  #
  # Note that the report dates are once a week, and reference dates are daily
  data <- data.table(
    report_date = as.Date(c(
      "2022-10-25", "2022-11-01", "2022-11-01", "2022-11-01", "2022-11-01", "2022-11-01",
      "2022-11-01", "2022-11-01", "2022-11-01", "2022-11-08", "2022-11-08", "2022-11-08",
      "2022-11-08", "2022-11-08", "2022-11-08", "2022-11-08", "2022-11-08", "2022-11-08",
      "2022-11-08", "2022-11-08"
    )),
    reference_date = as.Date(c(
      "2022-10-22", "2022-10-22", "2022-10-23", "2022-10-24", "2022-10-25", "2022-10-26",
      "2022-10-27", "2022-10-28", "2022-10-29", "2022-10-22", "2022-10-23", "2022-10-24",
      "2022-10-25", "2022-10-26", "2022-10-27", "2022-10-28", "2022-10-29", "2022-10-30",
      "2022-10-31", "2022-11-01"
    )),
    confirm = c(
      34, 46, 47, 41, 68, 59, 62, 30, 40,
      48, 53, 46, 75, 67, 84, 47, 67, 69, 81, 88
    )
  )

  # Want everything at a weekly timestep, so use
  # `data |> enw_aggregate_cumulative(timestep = "week")`
  # But this produces the error
  # Error in check_numeric_timestep(dates, date_var, internal_timestep, exact) :
  #   report_date does not have the specified timestep of 1 day(s)

  # Instead, use `data |> enw_complete_dates(timestep = "day")` to get a
  # consistent timestep of 1 day
  daily <- data |> enw_complete_dates(timestep = "day")

  # Complete the final step and aggregate to weekly timesteps for both reference
  # and report dates
  daily |> enw_aggregate_cumulative(timestep = "week")
  # Which gives the following table
  #    report_date reference_date confirm
  #         <IDat>         <IDat>   <num>
  # 1:  2022-10-29           <NA>       0
  # 2:  2022-11-05           <NA>       0
  # 3:  2022-10-29     2022-10-29       0
  # 4:  2022-11-05     2022-10-29     347
  # 5:  2022-11-05     2022-11-05       0

  # Note that even though we had reports before 2022-10-29, the report here for
  # 2022-10-29 with reference date of 2022-10-29 has value 0.
}

```

**Expected behavior**
I believe the row with report/reference pair 2022-10-29 should have a value of 34, from the very first row in `data` which has `report_date=2022-10-25`, `reference_date=2022-10-22`, `confirm=34`.

**Additional Context**
I installed the latest version of epinowcast with `pak::pak("epinowcast/epinowcast")`, and got the same results.

**Comment by pearsonca** [comment]
having been somewhat responsible for date munging, i'd bet this is my fault somehow.

@seabbs willing to let this sit til friday-ish? i have a long plane filght to be distracted from.

**Comment by natemcintosh** [comment]
Looks like it's dropping that very first row of the example data at [this line](https://github.com/epinowcast/epinowcast/blob/bd4ca01236a3bf7d94e33a18b747214949b439ef/R/data-converters.R#L356C3-L356C3) where `agg_obs` is created.

Skipping that line of code, and running all the others, I get the following output from `enw_aggregate_cumulative(timestep = "week")`
```r
   report_date reference_date confirm
        <IDat>         <IDat>   <num>
1:  2022-10-22           <NA>       0
2:  2022-10-29           <NA>       0
3:  2022-11-05           <NA>       0
4:  2022-10-22     2022-10-22       0
5:  2022-10-29     2022-10-22      34
6:  2022-11-05     2022-10-22      46
7:  2022-10-29     2022-10-29     204
8:  2022-11-05     2022-10-29    1713
9:  2022-11-05     2022-11-05     992
```

Not sure if this is actually any help since I don't know the purpose of that line of code.

**Comment by seabbs** [comment]
Thanks for reporting. I've fleshed out the reprex more fully with some expected outcomes and made the title of the issue clearer.

``` r
  library(data.table)
  library(epinowcast)

  # Edge case: Daily reference data with weekly report data where
  # the initial reference date is not a single timestep away from the
  # initial report date
  data <- data.table(
    report_date = as.Date(c(
      "2022-10-25", "2022-11-01", "2022-11-01", "2022-11-01", "2022-11-01",
      "2022-11-01", "2022-11-01", "2022-11-01", "2022-11-01", "2022-11-08",
      "2022-11-08", "2022-11-08", "2022-11-08", "2022-11-08", "2022-11-08",
      "2022-11-08", "2022-11-08", "2022-11-08", "2022-11-08", "2022-11-08"
    )),
    reference_date = as.Date(c(
      "2022-10-22", "2022-10-22", "2022-10-23", "2022-10-24", "2022-10-25",
      "2022-10-26", "2022-10-27", "2022-10-28", "2022-10-29", "2022-10-22",
      "2022-10-23", "2022-10-24", "2022-10-25", "2022-10-26", "2022-10-27",
      "2022-10-28", "2022-10-29", "2022-10-30", "2022-10-31", "2022-11-01"
    )),
    confirm = c(
      34, 46, 47, 41, 68, 59, 62, 30, 40,
      48, 53, 46, 75, 67, 84, 47, 67, 69, 81, 88
    )
  )

  # Get a complete daily timeseries as this is the basic format the package is
  # designed for
  daily <- data |>
    enw_complete_dates(timestep = "day", missing_reference = FALSE)

  # Complete the final step and aggregate to weekly timesteps for both reference
  # and report dates
  daily |>
    enw_aggregate_cumulative(timestep = "week")
#>    report_date reference_date confirm
#> 1:  2022-10-29     2022-10-29       0
#> 2:  2022-11-05     2022-10-29     347
#> 3:  2022-11-05     2022-11-05       0

  # Expected output
  ## Option 1: Incomplete weeks are dropped and the already weekly report date
  ## timestep is used
  weekly_data_by_report_date <- data.table(
    report_date = as.Date(
      c(
        "2022-10-25", "2022-11-01", "2022-11-08", "2022-11-01", "2022-11-08",
        "2022-11-08"
      )
    ),
    reference_date = as.Date(
      c(
        "2022-10-25", "2022-10-25", "2022-10-25", "2022-11-01", "2022-11-01",
        "2022-11-08"
      )
    ),
    confirm = c(34, 202, 222, 191, 503, 0)
  )
# Keep the first week of partially reported data
weekly_data_by_report_date[]
#>    report_date reference_date confirm
#> 1:  2022-10-25     2022-10-25      34
#> 2:  2022-11-01     2022-10-25     202
#> 3:  2022-11-08     2022-10-25     222
#> 4:  2022-11-01     2022-11-01     191
#> 5:  2022-11-08     2022-11-01     503
#> 6:  2022-11-08     2022-11-08       0

# Removing the first week of data that is partially reported
weekly_data_by_report_date[
  report_date != as.Date("2022-10-25")][reference_date != as.Date("2022-10-25")
]
#>    report_date reference_date confirm
#> 1:  2022-11-01     2022-11-01     191
#> 2:  2022-11-08     2022-11-01     503
#> 3:  2022-11-08     2022-11-08       0

## Option 2: A timestep based on the reference data is used and so there are no
## incomplete weeks
weekly_data_by_reference_date <- data.table(
  report_date = as.Date(c(
    "2022-10-28", "2022-11-04", "2022-11-04"
  )),
  reference_date = as.Date(c(
    "2022-10-28", "2022-10-28", "2022-11-04"

  )),
  confirm = c(34, 353, 40)
)
# Keep the first week of partially reported data
weekly_data_by_reference_date[]
#>    report_date reference_date confirm
#> 1:  2022-10-28     2022-10-28      34
#> 2:  2022-11-04     2022-10-28     353
#> 3:  2022-11-04     2022-11-04      40

# Don't keep the first week of partially reported data
weekly_data_by_reference_date[report_date != as.Date("2022-10-28")]
#>    report_date reference_date confirm
#> 1:  2022-11-04     2022-10-28     353
#> 2:  2022-11-04     2022-11-04      40
```

<sup>Created on 2023-10-18 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

So the expected behaviour is I believe option 2 and is not happening because in 

https://github.com/epinowcast/epinowcast/blob/bd4ca01236a3bf7d94e33a18b747214949b439ef/R/data-converters.R#L357

the minimum report date is incremented by one day (that line should include a minus one). The line more generally is there to prevent partially reported time steps (such as your first being included as these will be very problematic for modelling (as they aren't a true timestep). 

If you instead wanted option 1 like behaviour then that is outside the expected behaviour. You might want this because as here you do actually have a known timestep that you could be telling the package about vs allowing it to try and discover one. I'd suggest we open a new issue if this is the case to address this.

You could also potentially want the incomplete initial and final weeks but as above I think in virtually all cases that is very dangerous and if this is the case I think it might be on the user side to provide (as for most users this will just lead to errors).

**Comment by seabbs** [comment]
I've pushed a fix for the indexing off by 1 issue in #336 and suggest opening another issue if you would like to be able to specify a timestep start date (i.e. here that would be the 1st of November).
